packages/@core.keystone/errors.js

```diff
@@ -0,0 +1,128 @@
+ *
+ *                 // execute some another actions, that is not guaranteed to be succeed
+ *                 try {
+ *                     executeSomeAnotherAction
+ *                 } catch(e) {
+ *                     throw new GQLError({
+ *                         ...errors.UNABLE_EXECUTE_SOME_PROCEDURE,
+ *                         internalError: e
+ *                     })
+ *                 }
+ *             }
+ *         }
+ *     ]
+ * })
+ * ```
+ */
+const { ApolloError } = require('apollo-server-errors')
+
+// Unable to find a record, whose identifier is specified in some argument of query or mutation
+const NOT_FOUND = 'NOT_FOUND'
+// Data, being submitted, conflicts with constraints of a data storage
+// For example, it can violate constraints on uniqueness, number of records, formats of field values, values itself etc.
+const CONFLICT = 'CONFLICT'
+// Wrong format or not enough data in user input
+const BAD_USER_INPUT = 'BAD_USER_INPUT'
+// Generic error, that something went wrong at server side, though user input was correct
+const INTERNAL_ERROR = 'INTERNAL_ERROR'
+// Access denied. Semantically it corresponds to 403 HTTP code
+const FORBIDDEN = 'FORBIDDEN'
+
+
+/**
+ * First level of error classification, used in custom GraphQL queries or mutations
+ * Only generic error kinds are listed
+ * Conceptually, it conforms to HTTP standard for error codes
+ * @readonly
+ * @enum {String}
+ */
+const GQLErrorCode = {
+    NOT_FOUND,
+    CONFLICT,
+    BAD_USER_INPUT,
+    INTERNAL_ERROR,
+    FORBIDDEN,
+}
+
+// A record with specified set of field values already exists, so, the request violates unique constraints
+const NOT_UNIQUE = 'NOT_UNIQUE'
+
+// Second level of error classification
+const GQLErrorType = {
+    NOT_UNIQUE,
+}
+
+/**
+ * Error object, that can be thrown in a custom GraphQL mutation or query
+ *
+ * @typedef GQLError
+ * @property {String} [mutation] - name of mutation where the error has been occured
+ * @property {String} [query] - name of query where the error has been occured
+ * @property {Array.<String>} [variable] - path to mutation or query argument, that is a subject of an error
+ * @property {GQLErrorCode} code - standardized error code
+ * @property {String} message - humanized and error description that in future will be localized
+ * @property {String} [correctExample] - correct value of an argument
+ * @property {Object} [internalError] - error from internal part of the system. Not required, because in some cases it is not secure to expose internal error messages
+ */
+
+class GQLError extends ApolloError {
+    /**
+     * @param {GQLError} fields
+     * @see https://www.apollographql.com/docs/apollo-server/data/errors/#custom-errors
+     */
+    constructor (fields) {
+        super(fields.message, fields.code, fields)
+        Object.defineProperty(this, 'name', { value: 'GraphQLError' })
+    }
+}
+
+module.exports = {
+    GQLError,
+    GQLErrorCode,
+    GQLErrorType,
```

Comment:
You don't need to put the `GQLErrorType` here. It's a part of app domain logic. Not a package stuff

Reply:
`NOT_UNIQUE`, `WRONG_FORMAT` are general-purpose codes.
Btw, it seems strange to me to implement all of this in `@core.keystone` package.
Logically, it is an errors module for our GraphQL API, not a keystone.